From 1a12531755255812b44d7983d3fb8111095c3d2e Mon Sep 17 00:00:00 2001
From: Alexandru Croitor <alexandru.croitor@qt.io>
Date: Mon, 30 May 2022 10:18:02 +0200
Subject: qmlimportscanner: Ensure deterministic output

QQmlDirParser::components and other methods return QMultiHash types
which don't have a predefined order.
Use QHashSeed::setDeterministicGlobalSeed() to always use the same
order for QHash keys.

Also sort the components and scripts for each module.

Task-number: QTBUG-103187
Change-Id: Iff7c1cd14a723a5aff7c8772a2b94b110fb6ca44
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
(cherry picked from commit fd79c5bc2676cd223feee8906e843893eba33b8b)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 tools/qmlimportscanner/main.cpp | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git x/qtdeclarative/tools/qmlimportscanner/main.cpp y/qtdeclarative/tools/qmlimportscanner/main.cpp
index 605ea03400..336abf188f 100644
--- x/qtdeclarative/tools/qmlimportscanner/main.cpp
+++ y/qtdeclarative/tools/qmlimportscanner/main.cpp
@@ -249,10 +249,14 @@ QVariantMap pluginsForModulePath(const QString &modulePath, const QString &versi
 
     if (!importsAndDependencies.isEmpty())
         pluginInfo[dependenciesLiteral()] = importsAndDependencies;
-    if (!componentFiles.isEmpty())
+    if (!componentFiles.isEmpty()) {
+        componentFiles.sort();
         pluginInfo[componentsLiteral()] = componentFiles;
-    if (!scriptFiles.isEmpty())
+    }
+    if (!scriptFiles.isEmpty()) {
+        scriptFiles.sort();
         pluginInfo[scriptsLiteral()] = scriptFiles;
+    }
 
     if (!parser.preferredPath().isEmpty())
         pluginInfo[preferLiteral()] = parser.preferredPath();
@@ -685,6 +689,9 @@ int main(int argc, char *argv[])
         return 1;
     }
 
+    // QQmlDirParser returnes QMultiHashes. Ensure deterministic output.
+    QHashSeed::setDeterministicGlobalSeed();
+
     QStringList qmlRootPaths;
     QStringList scanFiles;
     QStringList qmlImportPaths;
-- 
2.37.1

